### ISSUE TITLE
Garbage in projectile.projects resulting in "wrong-type-argument listp nil/\.doom\.d/"

### ISSUE THREAD

**Issue by czikus** [hlissner/doom-emacs#4226]
**What actually happened?**

Roughly several times a day, Emacs starts misbehaving resulting in the following backtrace for many commands (examples for `save-buffers-kill-emacs` or magit below). After observing this problem for at least a month, I am still unable to tell what causes the problem or reliably reproduce it. However, when it happens `(projectile-unserialize projectile-known-projects-file)` indeed produces  `nil/\.doom\.d/` and the `projectile.projects` file contains `nil/.doom.d/")`. To resolve the problem, I have to run `(projectile-save-known-projects)`, which will overwrite the `projectile.projects` file. Often, the list of projects is lost in the process.

```
Debugger entered--Lisp error: (wrong-type-argument listp nil/\.doom\.d/)
  (#f(compiled-function (x) #<bytecode 0x1568b0e9c6f1>) "~/.doom.d/")
  (cl-remove nil ("~/.doom.d/") :if #f(compiled-function (x) #<bytecode 0x1568b0e9c6f1>))
  (apply cl-remove nil ("~/.doom.d/") :if #f(compiled-function (x) #<bytecode 0x1568b0e9c6f1>) nil)
  (cl-remove-if #f(compiled-function (x) #<bytecode 0x1568b0e9c6f1>) ("~/.doom.d/"))
  (projectile-difference ("~/.doom.d/") nil/\.doom\.d/)
  (projectile-merge-known-projects)
  (projectile--cleanup-known-projects)
  (projectile-cleanup-known-projects)
  (progn (projectile-cleanup-known-projects) (let* ((blacklist (mapcar #'file-truename doom-projectile-cache-blacklist)) (--cl-var-- (hash-table-keys projectile-projects-cache)) (proot nil)) (while (consp --cl-var--) (setq proot (car --cl-var--)) (if (or (not (stringp proot)) (>= (length (gethash proot projectile-projects-cache)) doom-projectile-cache-limit) (member (substring proot 0 -1) blacklist) (and doom-projectile-cache-purge-non-projects (not (doom-project-p proot)))) (progn (if doom-debug-p (progn (let ... ...))) (remhash proot projectile-projects-cache) (remhash proot projectile-projects-cache-time) (remhash proot projectile-project-type-cache))) (setq --cl-var-- (cdr --cl-var--))) nil) (projectile-serialize-cache))
  (if (and (and (boundp 'projectile-projects-cache) projectile-projects-cache) projectile-enable-caching doom-interactive-p) (progn (projectile-cleanup-known-projects) (let* ((blacklist (mapcar #'file-truename doom-projectile-cache-blacklist)) (--cl-var-- (hash-table-keys projectile-projects-cache)) (proot nil)) (while (consp --cl-var--) (setq proot (car --cl-var--)) (if (or (not (stringp proot)) (>= (length ...) doom-projectile-cache-limit) (member (substring proot 0 -1) blacklist) (and doom-projectile-cache-purge-non-projects (not ...))) (progn (if doom-debug-p (progn ...)) (remhash proot projectile-projects-cache) (remhash proot projectile-projects-cache-time) (remhash proot projectile-project-type-cache))) (setq --cl-var-- (cdr --cl-var--))) nil) (projectile-serialize-cache)))
  (doom-cleanup-project-cache-h)
  (kill-emacs)
  (save-buffers-kill-emacs nil)
  (#<subr funcall-interactively> save-buffers-kill-emacs nil)
  (apply #<subr funcall-interactively> (save-buffers-kill-emacs nil))
  (funcall-interactively save-buffers-kill-emacs nil)
  (#<subr call-interactively> save-buffers-kill-emacs nil nil)
  (apply #<subr call-interactively> (save-buffers-kill-emacs nil nil))
  (explain-pause--wrap-call-interactively #<subr call-interactively> save-buffers-kill-emacs nil nil)
  (apply explain-pause--wrap-call-interactively #<subr call-interactively> (save-buffers-kill-emacs nil nil))
  (call-interactively save-buffers-kill-emacs nil nil)
  (command-execute save-buffers-kill-emacs)
```

```
Debugger entered--Lisp error: (wrong-type-argument listp nil/\.doom\.d/)
  (#f(compiled-function (x) #<bytecode 0x1568b1c7b48d>) "~/.doom.d/")
  (cl-remove nil ("~/.doom.d/") :if #f(compiled-function (x) #<bytecode 0x1568b1c7b48d>))
  (apply cl-remove nil ("~/.doom.d/") :if #f(compiled-function (x) #<bytecode 0x1568b1c7b48d>) nil)
  (cl-remove-if #f(compiled-function (x) #<bytecode 0x1568b1c7b48d>) ("~/.doom.d/"))
  (projectile-difference ("~/.doom.d/") nil/\.doom\.d/)
  (projectile-merge-known-projects)
  (projectile-add-known-project "/home/user/.doom.d/")
  (projectile-track-known-projects-find-file-hook)
  (projectile-find-file-hook-function)
  (run-hooks find-file-hook)
  (#f(compiled-function (&optional error warn noauto after-find-file-from-revert-buffer nomodes) "Called after finding a file and by the default revert function.\nSets buffer mode, parses file-local and directory-local variables.\nOptional args ERROR, WARN, and NOAUTO: ERROR non-nil means there was an\nerror in reading the file.  WARN non-nil means warn if there\nexists an auto-save file more recent than the visited file.\nNOAUTO means don't mess with auto-save mode.\nFourth arg AFTER-FIND-FILE-FROM-REVERT-BUFFER is ignored\n(see `revert-buffer-in-progress-p' for similar functionality).\nFifth arg NOMODES non-nil means don't alter the file's modes.\nFinishes by calling the functions in `find-file-hook'\nunless NOMODES is non-nil." #<bytecode 0x1fc38577a6d3>) nil nil t nil nil)
  (apply #f(compiled-function (&optional error warn noauto after-find-file-from-revert-buffer nomodes) "Called after finding a file and by the default revert function.\nSets buffer mode, parses file-local and directory-local variables.\nOptional args ERROR, WARN, and NOAUTO: ERROR non-nil means there was an\nerror in reading the file.  WARN non-nil means warn if there\nexists an auto-save file more recent than the visited file.\nNOAUTO means don't mess with auto-save mode.\nFourth arg AFTER-FIND-FILE-FROM-REVERT-BUFFER is ignored\n(see `revert-buffer-in-progress-p' for similar functionality).\nFifth arg NOMODES non-nil means don't alter the file's modes.\nFinishes by calling the functions in `find-file-hook'\nunless NOMODES is non-nil." #<bytecode 0x1fc38577a6d3>) (nil nil t nil nil))
  (#f(advice-wrapper :before #f(compiled-function (&optional error warn noauto after-find-file-from-revert-buffer nomodes) "Called after finding a file and by the default revert function.\nSets buffer mode, parses file-local and directory-local variables.\nOptional args ERROR, WARN, and NOAUTO: ERROR non-nil means there was an\nerror in reading the file.  WARN non-nil means warn if there\nexists an auto-save file more recent than the visited file.\nNOAUTO means don't mess with auto-save mode.\nFourth arg AFTER-FIND-FILE-FROM-REVERT-BUFFER is ignored\n(see `revert-buffer-in-progress-p' for similar functionality).\nFifth arg NOMODES non-nil means don't alter the file's modes.\nFinishes by calling the functions in `find-file-hook'\nunless NOMODES is non-nil." #<bytecode 0x1fc38577a6d3>) doom-first-buffer-hook-h) nil nil t nil nil)
  (apply #f(advice-wrapper :before #f(compiled-function (&optional error warn noauto after-find-file-from-revert-buffer nomodes) "Called after finding a file and by the default revert function.\nSets buffer mode, parses file-local and directory-local variables.\nOptional args ERROR, WARN, and NOAUTO: ERROR non-nil means there was an\nerror in reading the file.  WARN non-nil means warn if there\nexists an auto-save file more recent than the visited file.\nNOAUTO means don't mess with auto-save mode.\nFourth arg AFTER-FIND-FILE-FROM-REVERT-BUFFER is ignored\n(see `revert-buffer-in-progress-p' for similar functionality).\nFifth arg NOMODES non-nil means don't alter the file's modes.\nFinishes by calling the functions in `find-file-hook'\nunless NOMODES is non-nil." #<bytecode 0x1fc38577a6d3>) doom-first-buffer-hook-h) (nil nil t nil nil))
  (#f(advice-wrapper :before #f(advice-wrapper :before #f(compiled-function (&optional error warn noauto after-find-file-from-revert-buffer nomodes) "Called after finding a file and by the default revert function.\nSets buffer mode, parses file-local and directory-local variables.\nOptional args ERROR, WARN, and NOAUTO: ERROR non-nil means there was an\nerror in reading the file.  WARN non-nil means warn if there\nexists an auto-save file more recent than the visited file.\nNOAUTO means don't mess with auto-save mode.\nFourth arg AFTER-FIND-FILE-FROM-REVERT-BUFFER is ignored\n(see `revert-buffer-in-progress-p' for similar functionality).\nFifth arg NOMODES non-nil means don't alter the file's modes.\nFinishes by calling the functions in `find-file-hook'\nunless NOMODES is non-nil." #<bytecode 0x1fc38577a6d3>) doom-first-buffer-hook-h) doom-first-file-hook-h) nil nil t nil nil)
  (apply #f(advice-wrapper :before #f(advice-wrapper :before #f(compiled-function (&optional error warn noauto after-find-file-from-revert-buffer nomodes) "Called after finding a file and by the default revert function.\nSets buffer mode, parses file-local and directory-local variables.\nOptional args ERROR, WARN, and NOAUTO: ERROR non-nil means there was an\nerror in reading the file.  WARN non-nil means warn if there\nexists an auto-save file more recent than the visited file.\nNOAUTO means don't mess with auto-save mode.\nFourth arg AFTER-FIND-FILE-FROM-REVERT-BUFFER is ignored\n(see `revert-buffer-in-progress-p' for similar functionality).\nFifth arg NOMODES non-nil means don't alter the file's modes.\nFinishes by calling the functions in `find-file-hook'\nunless NOMODES is non-nil." #<bytecode 0x1fc38577a6d3>) doom-first-buffer-hook-h) doom-first-file-hook-h) (nil nil t nil nil))
  (after-find-file nil nil t nil nil)
  (revert-buffer--default t t)
  (revert-buffer t t)
  (if (buffer-modified-p (current-buffer)) (if (and (boundp 'vc-mode) vc-mode) (progn (vc-refresh-state) (force-mode-line-update))) (revert-buffer t t))
  (progn (if (buffer-modified-p (current-buffer)) (if (and (boundp 'vc-mode) vc-mode) (progn (vc-refresh-state) (force-mode-line-update))) (revert-buffer t t)))
  (if buffer-file-name (progn (if (buffer-modified-p (current-buffer)) (if (and (boundp 'vc-mode) vc-mode) (progn (vc-refresh-state) (force-mode-line-update))) (revert-buffer t t))))
  (save-current-buffer (set-buffer buffer) (kill-local-variable '+magit--stale-p) (if buffer-file-name (progn (if (buffer-modified-p (current-buffer)) (if (and (boundp 'vc-mode) vc-mode) (progn (vc-refresh-state) (force-mode-line-update))) (revert-buffer t t)))))
  (+magit--revert-buffer #<buffer dockwin.el>)
  (if (get-buffer-window buffer) (+magit--revert-buffer buffer) (save-current-buffer (set-buffer buffer) (set (make-local-variable '+magit--stale-p) t)))
  (progn (if (get-buffer-window buffer) (+magit--revert-buffer buffer) (save-current-buffer (set-buffer buffer) (set (make-local-variable '+magit--stale-p) t))))
  (if (buffer-live-p buffer) (progn (if (get-buffer-window buffer) (+magit--revert-buffer buffer) (save-current-buffer (set-buffer buffer) (set (make-local-variable '+magit--stale-p) t)))))
  (let ((buffer (car --dolist-tail--))) (if (buffer-live-p buffer) (progn (if (get-buffer-window buffer) (+magit--revert-buffer buffer) (save-current-buffer (set-buffer buffer) (set (make-local-variable '+magit--stale-p) t))))) (setq --dolist-tail-- (cdr --dolist-tail--)))
  (while --dolist-tail-- (let ((buffer (car --dolist-tail--))) (if (buffer-live-p buffer) (progn (if (get-buffer-window buffer) (+magit--revert-buffer buffer) (save-current-buffer (set-buffer buffer) (set (make-local-variable ...) t))))) (setq --dolist-tail-- (cdr --dolist-tail--))))
  (let ((--dolist-tail-- (buffer-list))) (while --dolist-tail-- (let ((buffer (car --dolist-tail--))) (if (buffer-live-p buffer) (progn (if (get-buffer-window buffer) (+magit--revert-buffer buffer) (save-current-buffer (set-buffer buffer) (set ... t))))) (setq --dolist-tail-- (cdr --dolist-tail--)))))
  (+magit-mark-stale-buffers-h)
  (if (delq nil (mapcar #'(lambda (win) (let ((save-selected-window--state ...)) (save-current-buffer (unwind-protect ... ...)))) (window-list))) nil (mapc #'+magit--kill-buffer (magit-mode-get-buffers)) (+magit-mark-stale-buffers-h))
  (+magit/quit nil)
  (#<subr funcall-interactively> +magit/quit nil)
  (apply #<subr funcall-interactively> (+magit/quit nil))
  (funcall-interactively +magit/quit nil)
  (#<subr call-interactively> +magit/quit nil nil)
  (apply #<subr call-interactively> (+magit/quit nil nil))
  (explain-pause--wrap-call-interactively #<subr call-interactively> +magit/quit nil nil)
  (apply explain-pause--wrap-call-interactively #<subr call-interactively> (+magit/quit nil nil))
  (call-interactively +magit/quit nil nil)
  (command-execute +magit/quit)
```

**Steps to reproduce:**

Unable to reproduce, however the problem appears at random several times a day.

**System information:**

```
EMACS
        version 27.1
        build      Sep 08, 2020
        buildopts  --with-mailutils --with-kerberos5 --with-kerberos --with-json --with-cairo --with-imagemagick --with-x-toolkit=gtk3 --with-xwidgets
        features   XPM JPEG TIFF GIF PNG RSVG CAIRO IMAGEMAGICK SOUND GPM DBUS GSETTINGS GLIB NOTIFY INOTIFY ACL LIBSELINUX GNUTLS LIBXML2 FREETYPE HARFBUZZ M17N_FLT LIBOTF ZLIB TOOLKIT_SCROLL_BARS GTK3 X11 XDBE XIM MODULES THREADS XWIDGETS JSON PDUMPER LCMS2 GMP
        traits     (daemon server-running)
DOOM    dir        ~/.doom.d/
        version    2.0.9
        build      HEAD -> develop, origin/develop, origin/HEAD ee6c27c2a 2020-11-06 02:33:53 -0500
        elc-files  0
        modules    (:completion (company +tng) (helm +fuzzy) :ui doom doom-dashboard hl-todo modeline ophints vc-gutter (window-select +switch-window) workspaces :editor file-templates fold multiple-cursors snippets format :emacs dired electric (undo +tree) vc :checkers syntax (spell +hunspell +flyspell) :tools (eval +overlay) (lookup +dictionary +docsets) (lsp +peek) magit pdf rgb upload :lang emacs-lisp (org +pretty +dragndrop +journal +noter +roam) (markdown +grip) (sh +lsp +fish) (python +lsp +pyright +pyenv +poetry) (json +lsp) (javascript +lsp) (yaml +lsp) (web +lsp) (php +lsp) (cc +lsp) :config (default +bindings +smartparens))
        packages   ((i3wm-config-mode :recipe (:local-repo ~/.doom.d/packages/i3wm-config-mode :files (*.el))) (ssh-config-mode) (buffer-move) (aggressive-indent) (jump-char) (langtool)))
        unpin      (n/a)
        elpa       (n/a)
```

**Comment by hlissner** [comment]
I cannot reproduce this in vanilla Doom, much less make sense of this issue. This implies that `nil/.doom.d/` is being saved into your known projects somehow, despite projectile's safe-guards to prevent this. I'm inclined to suspect your private config or unconventional symlinking of your dotfiles. At the very least, evaluate `(projectile-project-root doom-private-dir)` and see if it returns a non-nil value.

If it does, then this needs more digging and taking projectile apart to determine the cause.

**Comment by czikus** [comment]
I'm quite puzzled by this myself. But I can definitely confirm that when the problem occurs `(projectile-project-root doom-private-dir)` evaluates to `"/home/user/.doom.d/"` and the folder is not symlinked.

**Comment by czikus** [comment]
I did some more digging. I added a message to `projectile-serialize` (`(message "Projectile writing to %s: %s" filename data)`) and `projectile-unserialize` (`(message "Projectile reading from %s: %s" filename (buffer-string))`) and got the following:

```
Projectile writing to /home/user/.emacs.d/.local/cache/projectile.projects: (~/.doom.d/)
Running git push -v origin refs/heads/master:refs/heads/master
Git finished
Projectile reading from /home/user/.emacs.d/.local/cache/projectile.projects: nil/.doom.d/")
```

So the data in `projectile-serialize` are correct, but the `(buffer-string)` in `projectile-unserialize` is not. And that happens exactly when magit is working. Now thinking about it, the problem pretty much always occurs shortly after I interact with magit. Any ideas?
